### DISCUSSION TITLE
Refactor entity versioning

### DISCUSSION THREAD

**Discussion by mschop** [shopware/platform#3161] [RFC / Proposal]
**Effort**: middle/high

### Description:
When data is written to the database through DAL, always a new version is created for entities, which are versioned.
This causes the database to accumulate data very quickly.

E.g. we had orders in the database with ~1000 versions saved (caused by some bug, but nevertheless 1000 versions).

Every update in the db produces a new entry in the table.

This has the following disadvantages:

- more difficult to understand for developers (especially new) -> many plugins are doing wrong DML queries directly in the database
- Increased database sizes (slow backups, more load regarding master->slave sync)
- Increased index sizes -> slows down performance / increases RAM requirements

Instead of creating a new version directly in the database table, a change set should be written to a generic interface.

Interface example:

```
interface VersioningStorageInterface
{
  public function addChangeset(array $changeset): void;
}
```

This interface could have multiple implementations. E.g.:
- MysqlVersioningStorage
- MongoDBVersioningStorage
- NullVersioningStorage -> completely deactivate functionality, if not needed
- much more

The changeset could then be saved in what ever format the implementation prefers. E.g. in the case of Mongo DB the changeset would be simply saved as an document.

### Benefits

- Better performance
- Simpler database structure / queries
- Better stability regarding bad programmed plugins
- Shops can opt-out of this feature, if it not needed for a shop

### Break strategy

As far as I know, the versions are not used anywhere. It is currently not possible to revert a change in the admin interface. (please correct me, if I am wrong).

So this change would not break code which uses the DAL.

Code which does not use the DAL would require refactoring.

**Comment by OliverSkroblin** [comment]
Hi @mschop, 

from your proposal it is not yet completely clear for us about which area you are talking about of the DAL versioning. 

Do you mean the process of writing the `version*` tables? Or about the realization of the primary keys / foreign keys and that when a new version of a record is to be created, that we duplicate the record and write it as a new entry into the database?

Outsourcing/Abstracting the storage (version tables) for the changeset sounds good for us. 👍 

The implementation regarding the primary keys and foreign keys are neccessary for some features from our side :/

Greetings Oliver

**Comment by mschop** [comment]
@OliverSkroblin 

The `version*`-tables would stay with slight modifications and would be used, if the MysqlVersioningStorage is used.

I am talking about the additional entries in e.g. the `order` table. If I understand you correctly, you are using those for some feature? For what features are you using those entries? What is the purpose of those entries?
The actual payload of a write is contained in `version_commit_data`. So why do you need the old versions in e.g. the `order` table? I never understood that, because there is not even a revert functionality in Shopware admin interface. And even if there was one, the additional entries in the tables are superfluous and the problem can be solved much simpler IMHO.

I would like to extend the upper interface a little bit:

```
class Change
{
  function __construct(
    public readonly User $user,
    public readonly array $payload,
  ) { }
}

interface VersioningStorageInterface
{
  public function add(Change $change): void;

  /**
   * @return Collection<Change>
   */
  public function get(string $entity, int $versionAmount): Collection
}
```

I hope, I made my point more clear :thinking: .

Best Regards
Martin

**Comment by OliverSkroblin** [comment]
Hey @mschop,

Please note that the technique of versioning in the DAL and this usage insidethe `order` module are two different things. 

The current implementation in the order module is based on the idea to have orders versioned permanently and to be able to restore them at any time. This would of course be possible via an interface like yours, but was simply implemented differently by the developers at the time. That's not what the "double entry" is for.

These additional version entries in the database are intended for something like campaign systems in the storefront, so that entities that are to be introduced with the campaign can be selected ahead of time. 
We already use this feature for our CMS Publisher plugin, which allows you to create CMS campaigns early and view them in the storefront. This kind publisher features will be extended in the future and would only be possible in a weakened form if we don't already have the different versions inside the storage.

But to come back to the Order Module: **YES** I understand your concerns and would also be happy if our storage is not so heavily loaded, for a function that does not offer much added value, or was partly not thought through to the end. **What do you think about rewriting the proposal to that effect?**

**Comment by keulinho** [comment]
Thx @mschop  for your proposal, but actually there was not much movement around this in the last year(s) and we don't expect that to change, therefore to be completely transparent we'll close this proposal for now.
